### ISSUE TITLE
~/.bashrc can break man page help in VS Code

### ISSUE THREAD

**Issue by EmilySeville7cfg** [bash-lsp/bash-language-server#619]
# Description

Documentation is broken for `declare` builtin:

![image](https://user-images.githubusercontent.com/42812113/207047305-731baa23-970a-4e02-bc6e-67881d50de66.png)

```bash
#!/usr/bin/env bash

declare
```

Code must be saved to reproduce this issue

# Environment

Extension version: v1.21.0
VS Code version: Version: 1.74.0
OS: Ubuntu MATE 22.04

**Comment by skovhus** [comment]
Unfortunately I can't reproduce this. Can you help me by running the following in your terminal and upload the `declare-dump` file here?

```sh
bash -c 'help declare' > declare-dump
```

<img width="533" alt="Screenshot 2022-12-12 at 19 03 26" src="https://user-images.githubusercontent.com/1260305/207120446-65cc90e5-1dc5-4f71-b77f-0f717141c649.png">

**Comment by skovhus** [comment]
And is the error the same for other builtins like `alias` or `cd`?

**Comment by skovhus** [comment]
The `something in colour` seems like something coming from your system somehow. 🤔

**Comment by EmilySeville7cfg** [comment]
I figure out why I see such behavior. In `~/.bashrc` I source another `.sh` script which ouputs colored output (for test purpose). But it's unexpected that `~/.bashrc` breaks help.

My current `~/.bashrc`:
```bash
#!/usr/bin/env bash

source base.sh

alias reload_config=". $HOME/.bashrc"

alias print="echo -en"
alias println="echo -e"

alias print_raw="echo -n"
alias println_raw="echo"

alias desktop="cd $HOME/Desktop"
alias documents="cd $HOME/Documents"
alias downloads="cd $HOME/Downloads"
alias music="cd $HOME/Music"
alias pictures="cd $HOME/Pictures"
alias templates="cd $HOME/Templates"
alias videos="cd $HOME/Videos"

get_config_version() {
    [[ -n "$1" ]] && return $FAIL

    println "1.0.2"
}


is_login() {
    shopt -q login_shell
}

is_interactive() {
    [[ $- == *i* ]]
}

get_program_name() {
    print "$0"
}

get_function_name() {
    print "${FUNCNAME[-1]}"
}

get_caller_function_name() {
    print "${FUNCNAME[-2]}"
}


is_integer() {
    [[ -n "$2" ]] && return $FAIL

    declare number="$1"
    [[ $number == ?([+-])[[:digit:]]+ ]]
}

is_float() {
    [[ -n "$2" ]] && return $FAIL

    declare number="$1"
    [[ $number == ?[+-][[:digit:]]+.[[:digit:]]+ ]]
}

get_random_integer() {
    [[ -n "$3" ]] && return $FAIL
    declare -i from="${1:-0}"
    declare -i to="${2:-10}"

    (( from > to )) && return $FAIL

    if (( from == to)); then
        print $from
    else
        print $(( RANDOM % (to - from) + from ))
    fi
}

get_random_float() {
    return $FAIL
}

get_raw_foreground_color() {
    [[ -n "$2" ]] && return $FAIL

    declare color_name="$1"
    declare -A color_name_associations=(
        [none]=0
        [black]=30
        [red]=31
        [green]=32
        [yellow]=33
        [blue]=34
        [magenta]=35
        [cyan]=36
        [light-gray]=37
        [gray]=90
        [light-red]=91
        [light-green]=92
        [light-yellow]=93
        [light-blue]=94
        [light-magenta]=95
        [light-cyan]=96
        [white]=97
    )

    [[ ! -v color_name_associations[$color_name] ]] && return $FAIL
    print_raw "\e[${color_name_associations[$color_name]}m"
}

get_raw_background_color() {
    [[ -n "$2" ]] && return $FAIL

    declare color_name="$1"
    declare -A color_name_associations=(
        [none]=0
        [black]=40
        [red]=41
        [green]=42
        [yellow]=43
        [blue]=44
        [magenta]=45
        [cyan]=46
        [light-gray]=47
        [gray]=100
        [light-red]=101
        [light-green]=102
        [light-yellow]=103
        [light-blue]=104
        [light-magenta]=105
        [light-cyan]=106
        [white]=107
    )

    [[ ! -v color_name_associations[$color_name] ]] && return $FAIL
    print_raw "\e[${color_name_associations[$color_name]}m"
}

get_foreground_color() {
    [[ -n "$2" ]] && return $FAIL

    declare color_name="$1"
    declare -A color_name_associations=(
        [none]=0
        [black]=30
        [red]=31
        [green]=32
        [yellow]=33
        [blue]=34
        [magenta]=35
        [cyan]=36
        [light-gray]=37
        [gray]=90
        [light-red]=91
        [light-green]=92
        [light-yellow]=93
        [light-blue]=94
        [light-magenta]=95
        [light-cyan]=96
        [white]=97
    )

    [[ ! -v color_name_associations[$color_name] ]] && return $FAIL
    print "\e[${color_name_associations[$color_name]}m"
}

get_background_color() {
    [[ -n "$2" ]] && return $FAIL

    declare color_name="$1"
    declare -A color_name_associations=(
        [none]=0
        [black]=40
        [red]=41
        [green]=42
        [yellow]=43
        [blue]=44
        [magenta]=45
        [cyan]=46
        [light-gray]=47
        [gray]=100
        [light-red]=101
        [light-green]=102
        [light-yellow]=103
        [light-blue]=104
        [light-magenta]=105
        [light-cyan]=106
        [white]=107
    )

    [[ ! -v color_name_associations[$color_name] ]] && return $FAIL
    print "\e[${color_name_associations[$color_name]}m"
}

alias raw_foreground="get_raw_foreground_color"
alias raw_background="get_raw_background_color"

alias foreground="get_foreground_color"
alias background="get_background_color"


show_info() {
    [[ -n "$3" ]] && return $FAIL

    declare function_name="$1"
    declare description="$2"

    println "$(get_foreground_color green)Info from $(get_foreground_color cyan)\`$function_name\`$(get_foreground_color green): $(get_foreground_color none)$description"
}

show_warning() {
    [[ -n "$3" ]] && return $FAIL

    declare function_name="$1"
    declare description="$2"

    println "$(get_foreground_color yellow)Warning from $(get_foreground_color cyan)\`$function_name\`$(get_foreground_color yellow): $(get_foreground_color none)$description"
}

show_error() {
    [[ -n "$3" ]] && return $FAIL

    declare function_name="$1"
    declare description="$2"

    println "$(get_foreground_color red)Error from $(get_foreground_color cyan)\`$function_name\`$(get_foreground_color red): $(get_foreground_color none)$description"
}


show_incorrect_amount_of_arguments_error() {
    declare function_name="$1"
    declare -i minimum_argument_count="$2"
    declare -i maximum_argument_count="$3"

    (( minimum_argument_count > maximum_argument_count )) && return $FAIL

    declare range=$minimum_argument_count
    if (( minimum_argument_count != maximum_argument_count )); then
        range="[$minimum_argument_count..$maximum_argument_count]"
    fi

    show_error "$function_name" "too many arguments, expected $range arguments"
}

show_not_implemented_error() {
    [[ -n "$2" ]] && return $FAIL

    declare function_name="$1"
    show_error "$function_name" "is not implemented yet"
}

show_script_usage_error() {
    [[ -n "$2" ]] && return $FAIL

    declare function_name="$1"
    show_error "$function_name" "is not permitted to be used in scripts"
}

show_permission_error() {
    [[ -n "$2" ]] && return $FAIL

    declare function_name="$1"
    show_error "$function_name" "permissions are broken to perform the intended action"
}


__setup_history() {
    [[ -n "$1" ]] && {
        show_too_many_arguments_error "$(get_function_name)" 0
        return $FAIL
    }
    is_interactive || show_script_usage_error "$(get_function_name)"

    shopt -s histappend
    HISTCONTROL=ignoreboth
    HISTSIZE=1000
    HISTFILESIZE=2000
}

__setup_window() {
    [[ -n "$1" ]] && {
        show_too_many_arguments_error "$(get_function_name)" 0
        return $FAIL
    }
    is_interactive || show_script_usage_error "$(get_function_name)"

    shopt -s checkwinsize
}

__setup_environment() {
    [[ -n "$1" ]] && {
        show_too_many_arguments_error "$(get_function_name)" 0
        return $FAIL
    }
    is_interactive || show_script_usage_error "$(get_function_name)"

    export PATH="$PATH:/home/emilyseville7cfg/.local/bin"
}


new_bash_script() {
    is_interactive || show_script_usage_error "$(get_function_name)"
    read -r -p "Program name (without extension): " script_name
    declare script_path="$HOME/Documents/mine/bash/$script_name.sh"

    if echo '#!/usr/bin/env bash

echo hello world' > "$script_path"; then
        show_info "$(get_function_name)" "script \`$script_path\` was created"
    else
        show_permission_error "$(get_function_name)"
    fi
}

new_python_script() {
    is_interactive || show_script_usage_error "$(get_function_name)"
    read -r -p "Program name (without extension): " script_name
    declare script_path="$HOME/Documents/mine/bash/$script_name.py"

    if echo '#!/usr/bin/env python3

print("hello world")' > "$script_path"; then
        show_info "$(get_function_name)" "script \`$script_path\` was created"
    else
        show_permission_error "$(get_function_name)"
    fi
}

new_java_script() {
    is_interactive || show_script_usage_error "$(get_function_name)"
    read -r -p "Program name (without extension): " script_name
    declare script_path="$HOME/Documents/mine/bash/$script_name.java"

    if echo '#!/usr/bin/env python3

public class Main {
  public static void main(String[] args) {
    System.out.println("hello world!");
  }
}' > "$script_path"; then
        show_info "$(get_function_name)" "script \`$script_path\` was created"
    else
        show_permission_error "$(get_function_name)"
    fi
}


is_interactive || return

__setup_history
__setup_window
__setup_environment

if [ -z "${debian_chroot:-}" ] && [ -r /etc/debian_chroot ]; then
    debian_chroot=$(cat /etc/debian_chroot)
fi

case "$TERM" in
    xterm-color|*-256color)
        is_colored_prompt=$(true)
    ;;
esac

if [[ $is_colored_prompt == $TRUE ]]; then
    PS1='${debian_chroot:+($debian_chroot)}\[$(get_foreground_color cyan)\]\u@\h\[$(get_foreground_color none)\]:\[$(get_foreground_color magenta)\]\w\[\e[00m\]\n\$ '
else
    PS1='${debian_chroot:+($debian_chroot)}\u@\h:\w\n\$ '
fi
unset color_prompt

case "$TERM" in
    xterm*|rxvt*)
        PS1="\[\e]0;${debian_chroot:+($debian_chroot)}\u@\h: \w\a\]$PS1"
    ;;
    *)
    ;;
esac

```

My current `~/base.sh`:

```bash
#!/usr/bin/env bash

[[ ! -v SUCCESS ]] && declare -i -r SUCCESS=0
[[ ! -v FAILURE ]] && declare -i -r FAILURE=1
[[ ! -v FAILURE_WITH_INVALID_OPTION ]] && declare -i -r FAILURE_WITH_INVALID_OPTION=1
[[ ! -v FAILURE_WITH_INVALID_ARGUMENT ]] && declare -i -r FAILURE_WITH_INVALID_ARGUMENT=1
[[ ! -v FAILURE_WITH_INVALID_ARGUMENT_ASSIGNMENT ]] && declare -i -r FAILURE_WITH_INVALID_ARGUMENT_ASSIGNMENT=1
[[ ! -v FAILURE_WITH_INVALID_ARGUMENT_OUT_OF_RANGE ]] && declare -i -r FAILURE_WITH_INVALID_ARGUMENT_OUT_OF_RANGE=1

[[ ! -v TRUE ]] && declare -i -r TRUE=$SUCCESS
[[ ! -v FALSE ]] && declare -i -r FALSE=$FAIL


true() {
  echo 0
  builtin true
}

false() {
  echo 1
  builtin false
}


function_name() {
  echo -n "${FUNCNAME[1]}"
}

caller_function_name() {
  echo -n "${FUNCNAME[2]}"
}

__core_function_failure_with_invalid_option_error() {
  echo "${1:-$(caller_function_name)}: wrong option passed, or too few or much options passed" >&2
}

__core_function_failure_with_invalid_argument_error() {
  echo "${1:-$(caller_function_name)}: wrong argument passed, or too few or much arguments passed" >&2
}

__core_function_failure_with_invalid_argument_assignment_error() {
  echo "${1:-$(caller_function_name)}: wrong argument assignment passed" >&2
}

__core_function_failure_with_invalid_argument_out_of_range_error() {
  if (( $# == 0 )); then
    echo "${1:-$(caller_function_name)}: wrong argument passed" >&2
  else
    echo "${1:-$(caller_function_name)}: wrong argument passed, expected one of: $*" >&2
  fi
}

__core_function_info() {
  [[ -n "$3" ]] && {
    __core_function_failure_with_invalid_argument_error "$2"
    return $FAILURE_WITH_INVALID_ARGUMENT
  }

  declare description="$1"

  echo "$(function_name): $(caller_function_name): $description"
}

__core_function_warn() {
  [[ -n "$3" ]] && {
    __core_function_failure_with_invalid_argument_error "$2"
    return $FAILURE_WITH_INVALID_ARGUMENT
  }

  declare description="$1"

  echo "$(function_name): $(caller_function_name): $description"
}

__core_function_error() {
  [[ -n "$3" ]] && {
    __core_function_failure_with_invalid_argument_error "$2"
    return $FAILURE_WITH_INVALID_ARGUMENT
  }

  declare description="$1"

  echo "$(function_name): $(caller_function_name): $description"
}


raw_foreground() {
  [[ -n "$2" ]] && {
    __core_function_failure_with_invalid_argument_error
    return $FAILURE_WITH_INVALID_ARGUMENT
  }

  declare color_name="$1"
  declare -A color_name_associations=(
    [none]=0
    [black]=30
    [red]=31
    [green]=32
    [yellow]=33
    [blue]=34
    [magenta]=35
    [cyan]=36
    [light-gray]=37
    [gray]=90
    [light-red]=91
    [light-green]=92
    [light-yellow]=93
    [light-blue]=94
    [light-magenta]=95
    [light-cyan]=96
    [white]=97
  )

  [[ ! -v color_name_associations[$color_name] ]] && {
    __core_function_failure_with_invalid_argument_out_of_range_error "${!color_name_associations[@]}"
    return $FAILURE_WITH_INVALID_ARGUMENT_OUT_OF_RANGE
  }
  echo -n "\e[${color_name_associations[$color_name]}m"
}

raw_background() {
  [[ -n "$2" ]] && {
    __core_function_failure_with_invalid_argument_error
    return $FAILURE_WITH_INVALID_ARGUMENT
  }

  declare -A color_name_associations=(
    [none]=0
    [black]=40
    [red]=41
    [green]=42
    [yellow]=43
    [blue]=44
    [magenta]=45
    [cyan]=46
    [light-gray]=47
    [gray]=100
    [light-red]=101
    [light-green]=102
    [light-yellow]=103
    [light-blue]=104
    [light-magenta]=105
    [light-cyan]=106
    [white]=107
  )

  declare color_name="$1"

  [[ ! -v color_name_associations[$color_name] ]] && {
    __core_function_failure_with_invalid_argument_out_of_range_error "${!color_name_associations[@]}"
    return $FAILURE_WITH_INVALID_ARGUMENT_OUT_OF_RANGE
  }
  echo -n "\e[${color_name_associations[$color_name]}m"
}

foreground() {
  declare raw_color
  raw_color="$(raw_foreground "$1")" || exit $?
  echo -e "$raw_color"
}

background() {
  declare raw_color
  raw_color="$(raw_background "$1")" || exit $?
  echo -e "$raw_color"
}


colorize() {
  [[ -n "$2" ]] && {
    __core_function_failure_with_invalid_argument_error
    return $FAILURE_WITH_INVALID_ARGUMENT
  }

  declare output=" $1 "
  output=${output// true / $(raw_foreground cyan)true$(raw_foreground none) }
  output=${output// false / $(raw_foreground cyan)false$(raw_foreground none) }
  output="$(awk "{{some code here}}" <<< "$output")"

  echo -n "${output:1:${#output} - 2}"
}

info() {
  declare arguments=("$@")
  declare output
  output="$(__core_function_info "${arguments[0]}" "info" "${arguments[@]:2}" 2>&1)"
  if (( $? != 0 )); then
    echo "$output"
    if [[ $output == *"too few or much"* ]]; then
      exit $FAILURE_WITH_INVALID_ARGUMENT
    else
      exit $FAILURE_WITH_INVALID_ARGUMENT_OUT_OF_RANGE
    fi
  fi

  output=${output/__core_function_info/$(raw_foreground magenta)$(caller_function_name)$(raw_foreground none)}
  output=${output//:/$(raw_foreground red):$(raw_foreground none)}
  output=${output/ info/ $(raw_foreground "$INFO_FOREGROUND_COLOR")$(raw_background "$INFO_BACKGROUND_COLOR")info$(raw_foreground none)}
  output="$(colorize "$output")"
  echo -e "$output"
}

warn() {
  declare arguments=("$@")
  declare output
  output="$(__core_function_warn "${arguments[0]}" "warn" "${arguments[@]:2}" 2>&1)"
  if (( $? != 0 )); then
    echo "$output"
    if [[ $output == *"too few or much"* ]]; then
      exit $FAILURE_WITH_INVALID_ARGUMENT
    else
      exit $FAILURE_WITH_INVALID_ARGUMENT_OUT_OF_RANGE
    fi
  fi

  output=${output/__core_function_warn/$(raw_foreground magenta)$(caller_function_name)$(raw_foreground none)}
  output=${output//:/$(raw_foreground red):$(raw_foreground none)}
  output=${output/ warn/ $(raw_foreground "$WARN_FOREGROUND_COLOR")$(raw_background "$WARN_BACKGROUND_COLOR")warn$(raw_foreground none)}
  output="$(colorize "$output")"
  echo -e "$output"
}

error() {
  declare arguments=("$@")
  declare output
  output="$(__core_function_error "${arguments[0]}" "error" "${arguments[@]:2}" 2>&1)"
  if (( $? != 0 )); then
    echo "$output"
    if [[ $output == *"too few or much"* ]]; then
      exit $FAILURE_WITH_INVALID_ARGUMENT
    else
      exit $FAILURE_WITH_INVALID_ARGUMENT_OUT_OF_RANGE
    fi
  fi

  output=${output/__core_function_error/$(raw_foreground magenta)$(caller_function_name)$(raw_foreground none)}
  output=${output//:/$(raw_foreground red):$(raw_foreground none)}
  output=${output/ error/ $(raw_foreground "$ERROR_FOREGROUND_COLOR")$(raw_background "$ERROR_BACKGROUND_COLOR")error$(raw_foreground none)}
  output="$(colorize "$output")"
  echo -e "$output"
}


existing_config_name() {
  declare config_name="$1"

  if [[ -e "$config_name.yml" ]]; then
    config_name+=".yml"
    echo "$config_name"
  elif [[ -e "$config_name.json" ]]; then
    config_name+=".json"
    echo "$config_name"
  else
    error "no '$config_name.yml' or '$config_name.json' file found"
    exit "$FAILURE_WITH_INVALID_ARGUMENT"
  fi
}

__core_function_load_message_color_defaults() {
  INFO_FOREGROUND_COLOR="${INFO_FOREGROUND_COLOR:-green}"
  WARN_FOREGROUND_COLOR="${WARN_FOREGROUND_COLOR:-yellow}"
  ERROR_FOREGROUND_COLOR="${ERROR_FOREGROUND_COLOR:-red}"

  INFO_BACKGROUND_COLOR="${INFO_BACKGROUND_COLOR:-white}"
  WARN_BACKGROUND_COLOR="${WARN_BACKGROUND_COLOR:-white}"
  ERROR_BACKGROUND_COLOR="${ERROR_BACKGROUND_COLOR:-white}"
}

__core_function_convert_message_colors_to_string() {
  echo -n "messages:
  info:
    foreground: $INFO_FOREGROUND_COLOR
    background: $INFO_BACKGROUND_COLOR
  warn:
    foreground: $WARN_FOREGROUND_COLOR
    background: $WARN_BACKGROUND_COLOR
  error:
    foreground: $ERROR_FOREGROUND_COLOR
    background: $ERROR_BACKGROUND_COLOR"
}

__core_function_load_config() {
  declare config_name="$HOME/bash-settings"

  if ! which yq > /dev/null; then
    __core_function_load_message_color_defaults
    warn "no 'yq' command found to load info, warn, error colors; defaults have been loaded:
  $(__core_function_convert_message_colors_to_string)"
  elif [[ ! -e "$config_name.yml" && ! -e "$config_name.json" ]]; then
    __core_function_load_message_color_defaults
    warn "no '$config_name.yml' or '$config_name.json' file found to load info, warn, error colors; defaults have been loaded:
  $(__core_function_convert_message_colors_to_string)"
  elif [[ -e "$config_name.yml" && -e "$config_name.json" ]]; then
    __core_function_load_message_color_defaults
    warn "both '$config_name.yml' and '$config_name.json' files found, leave one to load settings from; defaults have been loaded:
  $(__core_function_convert_message_colors_to_string)"
  else
    existing_config_name="$(existing_config_name "$HOME/bash-settings")"
    INFO_FOREGROUND_COLOR="$(yq '.messages.info.foreground // ""' "$existing_config_name")"
    WARN_FOREGROUND_COLOR="$(yq '.messages.warn.foreground // ""' "$existing_config_name")"
    ERROR_FOREGROUND_COLOR="$(yq '.messages.error.foreground // ""' "$existing_config_name")"

    INFO_BACKGROUND_COLOR="$(yq '.messages.info.background // ""' "$existing_config_name")"
    WARN_BACKGROUND_COLOR="$(yq '.messages.warn.background // ""' "$existing_config_name")"
    ERROR_BACKGROUND_COLOR="$(yq '.messages.error.background // ""' "$existing_config_name")"

    __core_function_load_message_color_defaults # Some value can be missing in config in which case default value must be used.
  fi
}

__core_function_load_config
exit

echo -e "$(awk "BEGIN{ print \"\\\e[35msomething in colour\\\e[0m\" }")"
echo -e "$(awk "BEGIN{ print \"\\$(raw_foreground magenta)something in colour\\$(raw_foreground none)\" }")"

awk "{ print gensub(/\"([^ ]+)\"/, \"\\\\1\", \"g\") }" <<< ' "some" '
awk -v start="\\\\\\$(raw_foreground magenta)" -v end="\\\\\\$(raw_foreground none)" "{ print gensub(/\"([^ ]+)\"/, \"\\\"\" + start + \"\\\\1\" + end + \"\\\"\", \"g\") }" <<< ' "some" '

gawk '{ print gensub(/(.+) (.+)/, "\\2 \\1", "g") }' <<< "abc def"
gawk "{ print gensub(/(.+) (.+)/, \"\\\\2 \\\\1\", \"g\") }" <<< "abc def"
#echo -e "$(awk "{ print gensub(/\"([^ ]+)\"/, \"\\$(raw_foreground magenta)\\1$(raw_foreground none)\", \"g\") }" <<< 'some text "with quoted" part and true boolean')"
#echo
#echo -e "$(colorize 'some text "with quoted" part and true boolean')"
#echo -e 'some text "with quoted" part and \e[36mtrue\e[0m boolean'
```

**Comment by EmilySeville7cfg** [comment]
@skovhus, I suppose `--noprofile` should be used then to ignore startup configs and fix this problem in this extension code.

**Comment by skovhus** [comment]
A fix is out, bash-language-server@4.1.1 and the vscode extension 1.22.0. Does it work for you?

**Comment by EmilySeville7cfg** [comment]
![image](https://user-images.githubusercontent.com/42812113/207526212-c4f17582-76dc-4fcf-9e7b-019d557ba71b.png)

Yes. 🚀 Thanks!
